### PR TITLE
python3 only

### PR DESCRIPTION
**Pull request by wdpypere**
- remove internal usage of py2vs3 module
 - remove colorized dependency
 - remove future module usage
 - happy new year
 - update tested python version
 - some resource decorators

needs some serious vetting

### CODE REVIEW

**Comment by wdpypere** [comment]
it could use way more cleanup, but this is big enough as it is to start. I would actually also just remove the daemon code, as it is unused afaik.

**Comment by wdpypere** [comment]
python 3.11 fails because Mapping moved places and I deleted the fix made in https://github.com/hpcugent/vsc-base/pull/335, as I do feel it does not belong there, as it is not an issue for py2v3. A new fix should be made.

**Comment by wdpypere** [comment]
as @wpoely86 also found out in #334 , python 3.11 and fancylogger are no friends. I'll look into it.

**Comment by wdpypere** [comment]
hmm, python 3.11 also does not like run_timeout

**Comment by wdpypere** [comment]
for run_timeout, now we don't need python2 support it seems to me we should switch from `subprocess.popen` to `subprocess.run`, it is more high level and supports timeout out of the box.

**Comment by wdpypere** [comment]
I would actually propose to add python 3.11 testing in a separate PR and have the discussion there.

File: .github/workflows/unittest.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ jobs:
       runs-on: ubuntu-20.04
       strategy:
         matrix:
-          python: [2.7, 3.6, 3.7, 3.8, 3.9]
+          python: [3.6, 3.9, 3.11]
```
Comment (wpoely86) [inline .github/workflows/unittest.yml:9]: Can you also pull in the other changes from #334?

Comment (wdpypere) [reply .github/workflows/unittest.yml:9]: @wpoely86 sure, but what do they do?

Comment (wpoely86) [reply .github/workflows/unittest.yml:9]: Fix deprecation warnings and use the newest version of all actions. And add python 3.10.

Comment (wdpypere) [reply .github/workflows/unittest.yml:9]: I was mainly wondering what the "newest version of all actions" means and implies.

File: .github/workflows/unittest.yml [hunk 2]
```diff
@@ -6,7 +6,7 @@ jobs:
       runs-on: ubuntu-20.04
```
Comment (stdweird) [inline .github/workflows/unittest.yml:6]: changes to this file should be generated with eg the `vsc.install.ci` tool. i propose to start with py3.6 and 3.9 only

File: lib/vsc/utils/fancylogger.py [hunk 3]
```diff
@@ -343,7 +323,7 @@ def raiseException(self, message, exception=None, catch=False):
             exception = self.RAISE_EXCEPTION_CLASS
 
         self.RAISE_EXCEPTION_LOG_METHOD(fullmessage)
-        raise_with_traceback(exception(message))
+        raise(exception(message))
```
Comment (stdweird) [inline lib/vsc/utils/fancylogger.py:326]: why is this removed/changed?

File: .github/workflows/unittest.yml (hunk 1 above)
Comment (wpoely86) [reply .github/workflows/unittest.yml:9]: You need to read the change notes for that :stuck_out_tongue: 

You currently have deprecation warnings. As seen in #334  it still works and does what was intended.

File: lib/vsc/utils/fancylogger.py [hunk 4]
```diff
@@ -559,10 +539,7 @@ def logToScreen(enable=True, handler=None, name=None, stdout=False, colorize=Col
     by default, logToScreen will log to stderr; logging to stdout instead can be done
     by setting the 'stdout' parameter to True
 
-    The `colorize` parameter enables or disables log colorization using
-    ANSI terminal escape sequences, according to the values allowed
-    in the `colorize` parameter to function `_screenLogFormatterFactory`
-    (which see).
+    The `colorize` is deprecated and does nothing.
```
Comment (stdweird) [inline lib/vsc/utils/fancylogger.py:542]: add warning when still used like the one below

File: lib/vsc/utils/frozendict.py [hunk 5]
```diff
@@ -22,13 +22,13 @@
...
 from functools import reduce
-from future.utils import iteritems
+try:
+    # py 3.10+
+    from collections import Mapping  # noqa
+except ImportError:
+    # < py 3.10
+    from collections.abc import Mapping  # noqa
 
-from vsc.utils.py2vs3 import Mapping
```
Comment (stdweird) [inline lib/vsc/utils/frozendict.py:27]: move this back to some module to handle all this crap somewhere
this will happen more, and it's really ugly to copy/paste these blocks

File: lib/vsc/utils/frozendict.py [hunk 6]
```diff
@@ -52,7 +52,7 @@ def __repr__(self):
 
     def __hash__(self):
         if self.__hash is None:
-            self.__hash = reduce(operator.xor, map(hash, iteritems(self)), 0)
+            self.__hash = reduce(operator.xor, map(hash, self.items()), 0)
```
Comment (stdweird) [inline lib/vsc/utils/frozendict.py:55]: are we sure this is the same?

File: lib/vsc/utils/missing.py [hunk 7]
```diff
@@ -44,10 +44,15 @@
 import time
 from collections import namedtuple
 from functools import reduce
+try:
```
Comment (stdweird) [inline lib/vsc/utils/missing.py:47]: import from something else

File: setup.py [hunk 8]
```diff
@@ -33,37 +33,23 @@
...
     'install_requires': [
         'vsc-install >= %s' % VSC_INSTALL_REQ_VERSION,
-        'future >= 0.16.0',
     ],
-    'extras_require': {
-        'coloredlogs': _coloredlogs_pkgs,
-    },
     'setup_requires': ['vsc-install >= %s' % VSC_INSTALL_REQ_VERSION],
-    'tests_require': ['prospector'] + _coloredlogs_pkgs,
+    'tests_require': ['prospector'],
```
Comment (stdweird) [inline setup.py:51]: this is odd. this should all come from vsc-install

File: .github/workflows/unittest.yml (hunk 2 above)
Comment (wdpypere) [reply .github/workflows/unittest.yml:6]: I agree, I'll make a ticket for vsc-install

File: lib/vsc/utils/frozendict.py (hunk 5 above)
Comment (wdpypere) [reply lib/vsc/utils/frozendict.py:27]: ah, I found out that since python 3.3 you can always do `from collections.abc import Mapping` so no more testing needed. I'll fix.

File: lib/vsc/utils/py2vs3/py2.py [hunk 9]
```diff
@@ -41,6 +42,7 @@
 FileExistsErrorExc = OSError  # noqa
 FileNotFoundErrorExc = OSError  # noqa
 
+logging.warning("py2 module deprecated, please stop using it.")
```
[the remark this answers is not in the document]
Comment (wdpypere) [reply lib/vsc/utils/py2vs3/py2.py:45]: I left it in because I did not remove the whole py2vs3 thing, I can however remove it, but it might break in other packages. I was thinking about removeing the whole py2vs3 in one go at a later stage, when more cleanups have happened.  I'll make it a raise.

[the remark this answers is not in the document]
Comment (wdpypere) [reply lib/vsc/utils/py2vs3/py2.py:45]: hmm, I might indeed just remove it and raise the exception in ```py2vs3/__init__.py```

File: setup.py (hunk 8 above)
Comment (wdpypere) [reply setup.py:51]: I'll try and remove it

File: .github/workflows/unittest.yml (hunk 2 above)
Comment (wdpypere) [reply .github/workflows/unittest.yml:6]: https://github.com/hpcugent/vsc-install/issues/214

File: lib/vsc/utils/fancylogger.py (hunk 3 above)
Comment (wdpypere) [reply lib/vsc/utils/fancylogger.py]: fixed

File: lib/vsc/utils/frozendict.py (hunk 5 above)
Comment (stdweird) [reply lib/vsc/utils/frozendict.py:27]: even better

File: lib/vsc/utils/frozendict.py (hunk 6 above)
Comment (wdpypere) [reply lib/vsc/utils/frozendict.py:50]: I'll doublecheck

Comment (wdpypere) [reply lib/vsc/utils/frozendict.py:50]:
```
>>> iteritems(thisdict)
dict_items([('brand', 'Ford'), ('model', 'Mustang'), ('year', 1964)])
>>> thisdict.items()
dict_items([('brand', 'Ford'), ('model', 'Mustang'), ('year', 1964)])
```

it does

**Comment by wdpypere** [comment]
@stdweird remarks addressed

File: .github/workflows/unittest.yml [hunk 10]
```diff
@@ -1,24 +1,28 @@
...
+      run: pip install tox
+    - name: add mandatory git remote
+      run: git remote add hpcugent https://github.com/hpcugent/vsc-base.git
+    - name: Run tox
+      run: tox -e py
+    strategy:
+      matrix:
+        python:
+        - 3.6
+        - 3.9
```
[the remark this answers is not in the document]
Comment (wdpypere) [reply .github/workflows/unittest.yml:24]: @itkovian these are the base pythons for RHEL8 and RHEL9. We know we'll have to fix stuff for 3.11, that's why it is not in here. if people use other pythons, they are welcome to open a PR and give a reasoning, but for now I know of no other case than 3.6 and 3.9 . (and 3.11 on fedora)

**Review by itkovian — APPROVED** [review]

File: lib/vsc/utils/fancylogger.py [hunk 11]
```diff
@@ -343,7 +323,7 @@ def raiseException(self, message, exception=None, catch=False):
             exception = self.RAISE_EXCEPTION_CLASS
 
         self.RAISE_EXCEPTION_LOG_METHOD(fullmessage)
-        raise_with_traceback(exception(message))
+        raise(exception(message)).with_traceback(tb)
```
Comment (stdweird) [inline lib/vsc/utils/fancylogger.py:326]: is this correct? `with_traceback` looks an exception method

so probably somethin g like
```
excp = exception(message)
if tb is not None:
    excp = excp.with_traceback(tb)
raise(excp)
```

File: test/fancylogger.py [hunk 12]
```diff
@@ -661,27 +630,27 @@ class ScreenLogFormatterFactoryTest(TestCase):
 
     def test_colorize_never(self):
```
Comment (stdweird) [inline test/fancylogger.py:631]: why do we keep the colorize tests? they are not supposed to do anything anymore, right?

File: test/run.py [hunk 13]
```diff
@@ -107,8 +106,8 @@ def test_startpath(self):
 
         # command was actually executed, in specified directory
         self.assertTrue(os.path.exists(os.path.join(self.tempdir, 'bar')))
-        txt = open(os.path.join(self.tempdir, 'bar')).read()
-        self.assertEqual(txt, 'foo\n')
+        with open(os.path.join(self.tempdir, 'bar')) as txt:
```
Comment (stdweird) [inline test/run.py:109]: use fp instead of txt

File: lib/vsc/utils/fancylogger.py (hunk 11 above)
Comment (wdpypere) [reply lib/vsc/utils/fancylogger.py:326]: well spotted, I'll fix.

**Comment by wdpypere** [comment]
@stdweird remarks addressed

File: lib/vsc/utils/fancylogger.py [hunk 14]
```diff
@@ -323,7 +323,9 @@ def raiseException(self, message, exception=None, catch=False):
             exception = self.RAISE_EXCEPTION_CLASS
 
         self.RAISE_EXCEPTION_LOG_METHOD(fullmessage)
-        raise(exception(message)).with_traceback(tb)
+        if tb is not None:
+            raise exception(message).with_traceback(tb)
```
Comment (stdweird) [inline lib/vsc/utils/fancylogger.py:327]: why the copy/paste? use a variable and build it up

Comment (wdpypere) [reply lib/vsc/utils/fancylogger.py:327]: it's a variable less. preference I guess. I'll change.

**Comment by wdpypere** [comment]
@stdweird remark fixed again :smile:
